### PR TITLE
ref(auth): Refactor auth_login.py

### PR DESCRIPTION
**Pull request by EricHasegawa**
Refactors `auth_login.py` and splits the mega `handle_basic_auth()` function into many smaller components.

**Why?** Auth is a consistently troublesome part of the codebase to work in. Seemingly small changes can have major side effects that are hard to predict and diagnose. Because auth code is brittle and hard to change, it's a limiting factor for projects that involve it (for example, the recent Fly.io work, or the updates we made to refresh tokens).  This change falls into a larger group of refactors/redesigns that the enterprise team is making in order to improve auth so we can:
- Ship code faster
- Add new functionality to Sentry auth without worrying so much about things breaking
- Lower the barrier to entry for working on auth code/flatten the related learning curve
- Change auth-related code confidently
- Avoid strange bugs in corner cases
- Trace errors and their causes more effectively

The goal here is to have the `auth_login.py` functionality be broken into small, predictable methods/functions with narrow scope that are easy to understand and easier to work with/change. The file is meant to be read like a top-down narrative, where, as you go deeper into the file, you descend layers of abstraction. Note that all of the main methods for `GET` come above the methods for `POST`, and then the shared methods/common utils are at the bottom.

Also note I left in the old `handle_basic_auth` function because some subclasses rely on it, after this I intend to switch those over and get rid of that too.

I know this is a sensitive file and I want to be careful in how we do this, if you notice any behavior that is untested, or can think of corner cases, please let me know and I can test for them.

### CODE REVIEW

File: src/sentry/web/frontend/auth_login.py [hunk 1]
```diff
@@ -95,39 +243,236 @@ def get_register_form(self, request: Request, initial=None):
...
+    ) -> None:
+        """Logs a user in and determines their active org."""
         login(request, user, organization_id=coerce_id_from(organization))
         self.determine_active_organization(request)
 
-    def handle_basic_auth(self, request: Request, **kwargs):
+    def refresh_organization_status(
+        self, request: Request, user: User, organization: Organization
+    ) -> None:
+        """I'm not really sure what this does actually."""
```
Comment (corps) [inline src/sentry/web/frontend/auth_login.py:405]: So the anwer to that is in the comment on he following line, and above on 384.

In general, organization context is determined at the start of our auth process with a ton of side effects.  In the case of completing login, this organization context is recalculated in a similar, *but not identical* way in order to inform a ton of other interactions that happen.

Yeah, it's a mess.  Tests do generally cover it however.

File: src/sentry/web/frontend/auth_login.py [hunk 2]
```diff
@@ -67,7 +67,134 @@
...
+
+    def get_sso_redirect(self, request: Request) -> HttpResponseRedirect:
+        """Returns a redirect response that will take a user to SSO login."""
+        url_prefix = generate_organization_url(request.subdomain)
+        url = absolute_uri(
+            reverse("sentry-auth-organization", args=[request.subdomain]), url_prefix=url_prefix
+        )
+        if request.GET:
+            url = f"{url}?{request.GET.urlencode()}"
+        return HttpResponseRedirect(url)
```
[the remark this answers is not in the document]
Comment (EricHasegawa) [reply src/sentry/web/frontend/auth_login.py]: These are reasonable detections - should we do anything about them?

File: src/sentry/web/frontend/auth_login.py [hunk 3]
```diff
@@ -95,39 +243,238 @@ def get_register_form(self, request: Request, initial=None):
...
+        self, request: Request, user: User, organization: Organization
+    ) -> None:
+        """Refresh organization status/context after a successful login to inform other interactions."""
+        # Refresh the organization we fetched prior to login in order to check its login state.
+        org_context = organization_service.get_organization_by_slug(
+            user_id=request.user.id,
+            slug=organization.slug,
+            only_visible=False,
+        )
+        if org_context:
```
Comment (nhsiehgit) [inline src/sentry/web/frontend/auth_login.py:414]: can we add some notes to what we're doing with the context here?

Looks like we're setting the active org on the session and/or stripping the 'next' redirect if the user is not an org member?

File: src/sentry/web/frontend/auth_login.py [hunk 4]
```diff
@@ -95,39 +243,238 @@ def get_register_form(self, request: Request, initial=None):
...
+            organization_service.check_organization_by_slug(
+                slug=request.subdomain, only_visible=True
+            )
+        )
 
-    def respond_login(self, request: Request, context, **kwargs):
-        return self.respond("sentry/login.html", context)
+    def can_register(self, request: Request) -> bool:
+        """Returns True if a user is eligible to register.
+        Users are eligible to register if they are arriving at this page via an invite link"""
```
Comment (nhsiehgit) [inline src/sentry/web/frontend/auth_login.py:440]: ahh that makes sense.

File: src/sentry/web/frontend/auth_login.py [hunk 5]
```diff
@@ -95,39 +243,238 @@ def get_register_form(self, request: Request, initial=None):
...
+        return bool(
+            organization_service.check_organization_by_slug(
+                slug=request.subdomain, only_visible=True
+            )
+        )
 
-    def respond_login(self, request: Request, context, **kwargs):
-        return self.respond("sentry/login.html", context)
+    def can_register(self, request: Request) -> bool:
+        """Returns True if a user is eligible to register.
```
Comment (nhsiehgit) [inline src/sentry/web/frontend/auth_login.py:439]: nit:

lets move the docstrings into their own line
```
"""
foobarbizbaz
"""
```

File: src/sentry/web/frontend/auth_login.py [hunk 6]
```diff
@@ -95,39 +243,238 @@ def get_register_form(self, request: Request, initial=None):
...
+        user.backend = settings.AUTHENTICATION_BACKENDS[0]
+        self._handle_login(request=request, user=user, organization=organization)
 
-        join_request_link_viewed.send_robust(sender=self, organization=organization)
+        # can_register should only allow a single registration
+        request.session.pop("can_register", None)
+        request.session.pop("invite_email", None)
 
-        return reverse("sentry-join-request", args=[organization.slug])
+        # Attempt to directly accept any pending invites
```
Comment (nhsiehgit) [inline src/sentry/web/frontend/auth_login.py:263]: All invite logic why don't we pull out into its own thing separate from a generic "redirec to next step" maybe?

File: src/sentry/web/frontend/auth_login.py [hunk 7]
```diff
@@ -82,11 +209,32 @@ def get_auth_provider(self, organization_slug):
 
         return auth_provider
 
-    def get_login_form(self, request: Request):
-        op = request.POST.get("op")
-        return AuthenticationForm(request, request.POST if op == "login" else None)
+    def handle_register_form_submit(
+        self, request: Request, context: dict, organization: Organization, **kwargs
+    ) -> HttpResponse:
+        """Validates a completed register form, redirecting to the next
```
Comment (nhsiehgit) [inline src/sentry/web/frontend/auth_login.py:215]: nit:
`"""`'s on their own line

Comment (nhsiehgit) [reply src/sentry/web/frontend/auth_login.py:215]: Can we also call out that "register" here means accepting invite link?

File: src/sentry/web/frontend/auth_login.py [hunk 8]
```diff
@@ -95,39 +243,238 @@ def get_register_form(self, request: Request, initial=None):
...
+    ) -> HttpResponseRedirect:
+        """Accepts an invite on behalf of a user and redirects them to their org login"""
+        invite_helper.accept_invite()
+        org_slug = invite_helper.invite_context.organization.slug
+        self.determine_active_organization(request=request, organization_slug=org_slug)
+        response = self.redirect_to_org(request=request)
+        remove_invite_details_from_session(request=request)
+        return response
+
+    def get_post_register_url(self, request: Request) -> str:
```
Comment (nhsiehgit) [inline src/sentry/web/frontend/auth_login.py:303]: oof 'post' here meaning "after" not `POST` 🙃 

can we think of a better word here?
'get_success_register_url'? eh...

File: src/sentry/web/frontend/auth_login.py [hunk 9]
```diff
@@ -95,39 +243,238 @@ def get_register_form(self, request: Request, initial=None):
...
+        """Gets the redirect URL for a successfully submitted register form."""
+        base_url = auth.get_login_redirect(request=request)
+        params = {"frontend_events": json.dumps({"event_name": "Sign Up"})}
+        return add_params_to_url(url=base_url, params=params)
 
-    def get_next_uri(self, request: Request):
+    def get_next_uri(self, request: Request) -> str:
```
Comment (nhsiehgit) [inline src/sentry/web/frontend/auth_login.py:309]: should this move up to the rest of the `GET` helpers?

**Review by nhsiehgit — COMMENTED** [review]
nice im loving it.

haven't gone through the tests, but have a bunch of nits/comments.

File: src/sentry/web/frontend/auth_login.py [hunk 10]
```diff
@@ -67,11 +67,138 @@ def run_callbacks(self, request: Request):
...
+        """Returns a redirect response that will take a user to SSO login."""
+        url_prefix = generate_organization_url(org_slug=request.subdomain)
+        url = absolute_uri(
+            url=reverse("sentry-auth-organization", args=[request.subdomain]), url_prefix=url_prefix
+        )
+        if request.GET:
+            url = f"{url}?{request.GET.urlencode()}"
+        return HttpResponseRedirect(redirect_to=url)
+
+    def get_registration_page(self, request: Request, context: dict, **kwargs) -> HttpResponse:
```
[the remark this answers is not in the document]
Comment (EricHasegawa) [reply src/sentry/web/frontend/auth_login.py:143]: They're returning the same page but they get different forms and update the `context` differently -> the differences in `context` make the page render registration stuff or login stuff. We definitely _could_ merge these into one method, but as a first step here I hope to have clear boundaries between things

File: src/sentry/web/frontend/auth_login.py [hunk 11]
```diff
@@ -67,11 +67,138 @@ def run_callbacks(self, request: Request):
...
+                request=request, context=default_context, organization=organization, **kwargs
+            )
+        else:
+            return self.handle_login_form_submit(
+                request=request, context=default_context, organization=organization
+            )
+
+    def redirect_post_to_sso(self, request: Request) -> HttpResponseRedirect:
+        """If the post call comes from the SSO tab, redirect the user to SSO login next steps."""
+        auth_provider = self.get_auth_provider_if_exists(org_slug=request.POST["organization"])
```
[the remark this answers is not in the document]
Comment (EricHasegawa) [reply src/sentry/web/frontend/auth_login.py:188]: agreed! there's a lot of behavior in here that I think we can and will improve - a lot of it comes to light when methods get separated out, however I think we should hold off on any actual changes to logic for another PR just so we can keep refactoring separate.

File: src/sentry/web/frontend/auth_login.py [hunk 12]
```diff
@@ -67,11 +67,138 @@ def run_callbacks(self, request: Request):
...
+        if auth_provider:
+            next_uri = reverse("sentry-auth-organization", args=[request.POST["organization"]])
+        else:
+            # Redirect to the org login route
+            next_uri = request.get_full_path()
+            messages.add_message(request=request, level=messages.ERROR, message=ERR_NO_SSO)
+
+        return HttpResponseRedirect(redirect_to=next_uri)
+
+    def get_auth_provider_if_exists(self, org_slug: str) -> Union[AuthProvider, None]:
```
[the remark this answers is not in the document]
Comment (EricHasegawa) [reply src/sentry/web/frontend/auth_login.py:198]: See above - we can do this in the next PR?

File: src/sentry/web/frontend/auth_login.py [hunk 13]
```diff
@@ -67,11 +67,138 @@ def run_callbacks(self, request: Request):
...
+        return response
+
+    def redirect_authenticated_user(self, request: Request, next_uri: str) -> HttpResponseRedirect:
+        """If an authenticated user sends a GET request to AuthLoginView, we redirect them forwards in the auth process."""
+        if is_valid_redirect(url=next_uri, allowed_hosts=(request.get_host(),)):
+            return self.redirect(url=next_uri)
+        return self.redirect_to_org(request=request)
+
+    def redirect_to_org_specific_login(self) -> HttpResponseRedirect:
+        """If Sentry is in single org mode this redirects users to their specific handler."""
```
[the remark this answers is not in the document]
Comment (EricHasegawa) [reply src/sentry/web/frontend/auth_login.py:117]: Yeah I'm pretty sure we can - would rather tackle that in a separate PR just so it's clear that this one isn't actually improving the "logic", it's just making it clear what that logic is

File: src/sentry/web/frontend/auth_login.py (hunk 7 above)
Comment (EricHasegawa) [reply src/sentry/web/frontend/auth_login.py:215]: Once you accept the invite link there's a form you have to fill out that confirms actual registration, so accepting the invite link is only one part of registration.

File: src/sentry/web/frontend/auth_login.py [hunk 14]
```diff
@@ -82,11 +209,32 @@ def get_auth_provider(self, organization_slug):
...
-    def get_login_form(self, request: Request):
-        op = request.POST.get("op")
-        return AuthenticationForm(request, request.POST if op == "login" else None)
+    def handle_register_form_submit(
+        self, request: Request, context: dict, organization: Organization, **kwargs
+    ) -> HttpResponse:
+        """Validates a completed register form, redirecting to the next
+        step or returning the form with its errors displayed."""
+
+        register_form = self.get_register_form(
```
[the remark this answers is not in the document]
Comment (EricHasegawa) [reply src/sentry/web/frontend/auth_login.py:218]: I like `get_` since it returns the form, "initialize" makes me think it's starting a form, but actually the form is already complete and it's just fetching it out of the request. We could name it "fetch" but that term has other connotations re. Javascripts Fetch and like fetching from a DB. I think `get` is a safe bet but I agree it _could_ be better, just not sure what would be better

File: src/sentry/web/frontend/auth_login.py [hunk 15]
```diff
@@ -95,39 +243,238 @@ def get_register_form(self, request: Request, initial=None):
...
+
+        context = {
+            "op": op or "login",
+            "login_form": login_form,
+        }
+
+        context.update(additional_context.run_callbacks(request))
+        return self.respond_login(request=request, context=context, **kwargs)
+
+    def redirect_to_next_login_step(
```
[the remark this answers is not in the document]
Comment (EricHasegawa) [reply src/sentry/web/frontend/auth_login.py:370]: `_handle_login` is a method from the old `handle_basic_auth` and it may have strange dependencies - i flattened out the logic here, but would prefer to get rid of `_handle_login` when we get rid out `handle_basic_auth` in the next PR.

File: src/sentry/web/frontend/auth_login.py (hunk 3 above)
Comment (EricHasegawa) [reply src/sentry/web/frontend/auth_login.py:414]: yea ill need to follow up with some more people to properly understand what this is doing, then ill document it better.

File: src/sentry/web/frontend/auth_login.py [hunk 16]
```diff
@@ -95,39 +243,238 @@ def get_register_form(self, request: Request, initial=None):
...
+
+                if org_member is None:
+                    org_member = organization_service.check_membership_by_id(
+                        organization_id=org_context.organization.id, user_id=user.id
+                    )
+                if org_member is None or org_member.user_id is None:
+                    request.session.pop("_next", None)
+
+    def check_if_org_exists(self, request: Request) -> bool:
+        """Returns True iff the organization passed in a request exists."""
```
[the remark this answers is not in the document]
Comment (EricHasegawa) [reply src/sentry/web/frontend/auth_login.py:431]: Yea i meant "if and only if" here, but i can switch it

File: src/sentry/web/frontend/auth_login.py [hunk 17]
```diff
@@ -67,11 +67,138 @@ def run_callbacks(self, request: Request):
...
+        if settings.SENTRY_SINGLE_ORGANIZATION:
+            return self.redirect_to_org_specific_login()
+
+        session_expired = "session_expired" in request.COOKIES
+        if session_expired:
+            messages.add_message(
+                request=request, level=messages.WARNING, message=WARN_SESSION_EXPIRED
+            )
+
+        default_context = self.get_default_context(request=request)
```
[the remark this answers is not in the document]
Comment (EricHasegawa) [reply src/sentry/web/frontend/auth_login.py:100]: Yes!

File: src/sentry/web/frontend/auth_login.py [hunk 18]
```diff
@@ -95,39 +243,238 @@ def get_register_form(self, request: Request, initial=None):
...
-            return None
+        user = register_form.save()
+        user.send_confirm_emails(is_new_user=True)
+        user_signup.send_robust(sender=self, user=user, source="register-form", referrer="in-app")
 
-        if organization.get_option("sentry:join_requests") is False:
-            return None
+        # HACK: grab whatever the first backend is and assume it works
+        user.backend = settings.AUTHENTICATION_BACKENDS[0]
+        self._handle_login(request=request, user=user, organization=organization)
```
[the remark this answers is not in the document]
Comment (EricHasegawa) [reply src/sentry/web/frontend/auth_login.py:295]: Yea good shout - fixed this up

File: src/sentry/web/frontend/auth_login.py [hunk 19]
```diff
@@ -45,12 +45,12 @@ class AdditionalContext:
     def __init__(self):
         self._callbacks = set()
 
-    def add_callback(self, callback):
+    def add_callback(self, callback: Any) -> None:
```
[the remark this answers is not in the document]
Comment (EricHasegawa) [reply src/sentry/web/frontend/auth_login.py:48]: I was trying to do this but I actually can't find anywhere where this is used - there's lots of `responses.add_callback()` in the codebase, but I can't find any using `AdditionalContext`. Do you know where this is used?

File: src/sentry/web/frontend/auth_login.py [hunk 20]
```diff
@@ -67,11 +69,168 @@ def run_callbacks(self, request: Request):
...
+
+    def redirect_authenticated_user(self, request: Request, next_uri: str) -> HttpResponseRedirect:
+        """
+        If an authenticated user sends a GET request to AuthLoginView, we redirect them forwards in the auth process.
+        """
+        if is_valid_redirect(url=next_uri, allowed_hosts=(request.get_host(),)):
+            return self.redirect(url=next_uri)
+        return self.redirect_to_org(request=request)
+
+    def redirect_to_org_specific_login(self) -> HttpResponseRedirect:
```
Comment (dashed) [inline src/sentry/web/frontend/auth_login.py:133]:
```suggestion
    def redirect_to_default_org_login(self) -> HttpResponseRedirect:
```

File: src/sentry/web/frontend/auth_login.py [hunk 21]
```diff
@@ -67,11 +69,168 @@ def run_callbacks(self, request: Request):
...
+        """
+        If Sentry is in single org mode this redirects users to their specific handler.
+        """
+        org = Organization.get_default()
```
Comment (dashed) [inline src/sentry/web/frontend/auth_login.py:137]: Queries to `Organization` should be done via `organization_service` as this view is not guaranteed to be called exclusively in the region silo.

```suggestion
        org = organization_service.get_default_organization()
```

File: src/sentry/web/frontend/auth_login.py [hunk 22]
```diff
@@ -67,11 +69,168 @@ def run_callbacks(self, request: Request):
...
+        if settings.SENTRY_SINGLE_ORGANIZATION:
+            return self.redirect_to_org_specific_login()
+
+        session_expired = "session_expired" in request.COOKIES
+        if session_expired:
+            messages.add_message(
+                request=request, level=messages.WARNING, message=WARN_SESSION_EXPIRED
+            )
+
+        default_context = self.get_default_context(request=request)
```
Comment (dashed) [inline src/sentry/web/frontend/auth_login.py:100]: `default_context = self.get_default_context(request=request)` can be moved into the functions `self.get_registration_page()` and `self.get_login_page()`.

File: src/sentry/web/frontend/auth_login.py [hunk 23]
```diff
@@ -67,11 +69,168 @@ def run_callbacks(self, request: Request):
...
+            }
+        )
+        return self.respond_login(request=request, context=context, **kwargs)
+
+    def post(self, request: Request, **kwargs) -> HttpResponse:
+        op = request.POST.get("op")
+        if op == "sso" and request.POST.get("organization"):
+            return self.redirect_post_to_sso(request=request)
+
+        default_context = self.get_default_context(request=request, **kwargs)
```
Comment (dashed) [inline src/sentry/web/frontend/auth_login.py:202]: `default_context = self.get_default_context(request=request)` can be moved into the functions `self.handle_register_form_submit()` and `self.handle_login_form_submit()`.

File: src/sentry/web/frontend/auth_login.py [hunk 24]
```diff
@@ -109,25 +531,14 @@ def get_join_request_link(self, organization):
...
-
-    def respond_login(self, request: Request, context, **kwargs):
-        return self.respond("sentry/login.html", context)
-
-    def _handle_login(self, request: Request, user, organization: Optional[Organization]):
-        login(request, user, organization_id=coerce_id_from(organization))
-        self.determine_active_organization(request)
-
-    def handle_basic_auth(self, request: Request, **kwargs):
+    def handle_basic_auth(
```
Comment (dashed) [inline src/sentry/web/frontend/auth_login.py:534]: Are you still keeping this method around? It doesn't seem to be used anymore.

File: src/sentry/web/frontend/auth_login.py [hunk 25]
```diff
@@ -67,11 +69,168 @@ def run_callbacks(self, request: Request):
...
+            reverse("sentry-auth-organization", args=[request.subdomain]),
+            reverse("sentry-register"),
+        ]
+        return (
+            request.subdomain
+            and self.org_exists(request=request)
+            and request.path_info not in non_sso_urls
+        )
+
+    def get_sso_redirect(self, request: Request) -> HttpResponseRedirect:
```
Comment (dashed) [inline src/sentry/web/frontend/auth_login.py:155]:
```suggestion
    def get_org_auth_login_redirect(self, request: Request) -> HttpResponseRedirect:
```
This isn't uniquely a SSO login redirect.

For example, both https://sentry.io/auth/login/ and https://sentry.io/auth/login/orgslug/ may be the same login template if `orgslug` hasn't set up any SSO provider. 

This just redirects to the login page specific for the organization (i.e. where the organization slug is fixed).

**Review by JoshFerge — COMMENTED** [review]
Overall, the changes look straightforward. Some comments on reducing risk:

1. It may be prudent to familiarize yourself with the various telemetry we have for login. Metrics dashboards, Sentry alerts, GCP logs, and so forth. If you don't know where these things are, let me know and I can direct you. It may also be worthwhile to add metrics / logs where necessary. Goal would be to make sure that when deployed, if there's an issue, you know before a customer has to report it.
2. I see in codecov, there is a function or two missing test coverage. would double check those cases and add if deemed important.
3. I'm curious how easy it is to have the new views use the new code instead of `handle_basic_auth`. it seems like it'd be pretty straightforward? might be worth doing locally at least, just to see if tests pass / those tests cover logic in that function that the base views don't. 
4. Aside from making the code more readable, it may be worth making UML diagrams describing the various sequences and use cases of these functions, and including links to those somewhere. 

good job! this was quite the undertaking, and the changes look clean to me.

File: src/sentry/web/frontend/auth_login.py (hunk 24 above)
Comment (EricHasegawa) [reply src/sentry/web/frontend/auth_login.py:534]: It's used in `auth_organization_login.py`, in the next PR I'll figure out what specifically it's doing and send it to the right function so we can get rid of `handle_basic_auth` soon. 

Since it's such a generalized function, `auth_organization_login` could basically just toss requests at it and the function would figure out what to do, now that functions are more specific, calls will need to be routed properly, which I'll do right after this

File: src/sentry/web/frontend/auth_login.py (hunk 21 above)
Comment (EricHasegawa) [reply src/sentry/web/frontend/auth_login.py]: Does this include things like `Organization.get()`? The previous code wasn't using this and I only see it being used in one place in the codebase - would it be better to swap that over in a follow-up PR in order to avoid scope-creep/any risk from making an actual change, or should we put it in here?

**Comment by EricHasegawa** [comment]
> Overall, the changes look straightforward. Some comments on reducing risk:
> 
> 1. It may be prudent to familiarize yourself with the various telemetry we have for login. Metrics dashboards, Sentry alerts, GCP logs, and so forth. If you don't know where these things are, let me know and I can direct you. It may also be worthwhile to add metrics / logs where necessary. Goal would be to make sure that when deployed, if there's an issue, you know before a customer has to report it.
> 2. I see in codecov, there is a function or two missing test coverage. would double check those cases and add if deemed important.
> 3. I'm curious how easy it is to have the new views use the new code instead of `handle_basic_auth`. it seems like it'd be pretty straightforward? might be worth doing locally at least, just to see if tests pass / those tests cover logic in that function that the base views don't.
> 4. Aside from making the code more readable, it may be worth making UML diagrams describing the various sequences and use cases of these functions, and including links to those somewhere.
> 
> good job! this was quite the undertaking, and the changes look clean to me.

Thank you! I'll address feedback one at a time:
1. I'd love to add logging to this, I'll look into it and see about a follow-up PR. One of the main advantages of the refactor is that stack traces / logs can be a lot more specific in how an error state was reached, so we should def take advantage of that.
2. All the untested functions are untested for a reason - if any of these are invalid or I'm mistaken please lmk:
   - `add_callback()` and the inner loop of `run_callbacks()` are untested because I actually can't find _anywhere_ in the codebase where these are used. `add_callback` never gets called as far as I can see, and the inner part of `run_callbacks` only gets triggered if callbacks have been added. After this is merged I might try deleting these and seeing if there's downstream impacts - maybe we can just get rid of them?
   - `refresh_organization_status()` will only ever actually get called when we enter `auth_login` from `auth_organization_login`, which currently still uses the legacy `handle_basic_auth`, so this function isn't really used yet (leaving it in this PR because I'll be switching `auth_organization_login` to use the new code soon).
   - Other than these, there are a couple untested edge cases (reactivating accounts) for example, that I do actually think should be covered due to the importance of this file, so I'll fix those up.
 3. What do you mean by "the new views"? Like `auth_organization_login`? If so, yes that will be fairly easy, basically I just need to find out where in the process to drop in their request and switch it over that way. I didn't include in this PR since it's already getting quite big and I don't want to increase risk.
 4. Great idea - at it's core this is a fairly understandable with a bunch of possible exit points, I can document this so people can pick it up faster in the future.

File: src/sentry/web/frontend/auth_login.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (nhsiehgit) [reply src/sentry/web/frontend/auth_login.py:225]: 👍 lets track in a ticket

File: src/sentry/web/frontend/auth_login.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (nhsiehgit) [reply src/sentry/web/frontend/auth_login.py:250]: 🤔 

That's not quite the case though.
This is constructing the form instance from the request - so in this case i _would_ say that we are initializing a form instance with data from the request.

Remember the request is just a normal xhr request. It doesn't contain any forms, it doesn't contain anything special to django at all. Just data that we're abstracting out to construct these new representations of that data.

File: src/sentry/web/frontend/auth_login.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (EricHasegawa) [reply src/sentry/web/frontend/auth_login.py:48]: https://getsentry.atlassian.net/jira/software/c/projects/ER/boards/90/backlog?view=detail&selectedIssue=ER-1733&issueLimit=100

File: src/sentry/web/frontend/auth_login.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (EricHasegawa) [reply src/sentry/web/frontend/auth_login.py:215]: https://getsentry.atlassian.net/jira/software/c/projects/ER/boards/90/backlog?view=detail&selectedIssue=ER-1733&issueLimit=100

File: src/sentry/web/frontend/auth_login.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (EricHasegawa) [reply src/sentry/web/frontend/auth_login.py:250]: ok im convinced. fixed!

**Review by nhsiehgit — COMMENTED** [review]
a handful more comments/questions/etc.

File: src/sentry/web/frontend/auth_login.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (EricHasegawa) [reply src/sentry/web/frontend/auth_login.py:48]: I'd prefer not to do that in this PR just in case it's required somewhere - it's in the follow-up Jira ticket!

File: src/sentry/web/frontend/auth_login.py [hunk 26]
```diff
@@ -67,11 +67,138 @@
...
+        """If the post call comes from the SSO tab, redirect the user to SSO login next steps."""
+        auth_provider = self.get_auth_provider_if_exists(org_slug=request.POST["organization"])
+        if auth_provider:
+            next_uri = reverse("sentry-auth-organization", args=[request.POST["organization"]])
+        else:
+            # Redirect to the org login route
+            next_uri = request.get_full_path()
+            messages.add_message(request=request, level=messages.ERROR, message=ERR_NO_SSO)
+
+        return HttpResponseRedirect(redirect_to=next_uri)
```
[the remark this answers is not in the document]
Comment (nhsiehgit) [reply src/sentry/web/frontend/auth_login.py:222]: warning suggests we validate the inputs first.
Any way we can add some validations for these?

File: src/sentry/web/frontend/auth_login.py [hunk 27]
```diff
@@ -82,23 +237,286 @@ def get_auth_provider(self, organization_slug):
...
+
+        return attempted_login and ratelimiter.is_limited(
+            "auth:login:username:{}".format(
+                md5_text(login_form.clean_username(value=request.POST["username"])).hexdigest()
+            ),
+            limit=5,
+            window=60,  # 5 per minute should be enough for anyone
+        )
+
+    def get_ratelimited_login_form(
```
Comment (nhsiehgit) [inline src/sentry/web/frontend/auth_login.py:401]: huh.
Do we need a completely separate helper for this?
Can we flatten this above as an optional arg in the generic return_login_form?

**Review by nhsiehgit — APPROVED** [review]
getting a little bit of review fatigue, but a few non blocking comments

Would like to tread carefully here.
If we can get at least 2 approvals here from different parties that would be ideal.

Once we start rolling this out, lets keep a careful eye on the [login stats](https://app.datadoghq.com/dashboard/ax8-c4b-rxd/login-signup?from_ts=1689350985983&to_ts=1689955785983&live=true)
possibly add more stats as well in this PR just to be safe so we can track properly/rollout safely

nice work 👍 🖖

File: src/sentry/web/frontend/auth_login.py (hunk 27 above)
Comment (EricHasegawa) [reply src/sentry/web/frontend/auth_login.py:401]: it'd just be copy pasting the 12-line helper into the above function, i prefer it to be separated in here so its clear whats happening, do you feel strongly about having it above?

**Review by dashed — APPROVED** [review]

**Review by JoshFerge — APPROVED** [review]
nice work. please make sure to familiarize yourself with GoCD and how to do a quick revert. may want to make sure dev infra is online and available to help in case.

I also assume you've done some basic sanity checks manually?

Additionally, it may be lower risk to deploy this _not_ in the morning, when everyone is usually logging on for the day.
